### PR TITLE
WIP: Node management

### PR DESCRIPTION
**Pull request by jackstar12**
Introduces a new Node Management UI, still in progress.
Only cln is implemented at the moment

### CODE REVIEW

**Comment by jackstar12** [comment]
current state of dashboard:
image

channels:
image

transactions are showing the payments that happened directly on the node:
image

**Comment by arcbtc** [comment]
Amazing. Can you remove "apps", thats for another pr

**Comment by arcbtc** [comment]
I love how it doesn't look boilerplate at all, just another lnbits menu item (for those who choose to enable). Very cool

**Comment by jackstar12** [comment]
been working on the channels tab:
image

**Comment by jackstar12** [comment]
I think I should clarify why supporting JSON columns in db would be beneficial.

I'll use the recently added `extra` column in the https://github.com/lnbits/usermanager (which currently doesnt work on sqlite!) as an example.

It is important that the usermanager api gives a way of filtering by custom data in the extra column.
There are essentially two ways to do this:
- fetch all users and filter application side
- filter directly in sql and only return required rows

The latter option is obviously much more performant, but it requires the column type to be json so that we can write queries accordingly. 
It also allows for easier partial updates. With a text column you'd have to select the row, do the update in python and then update. JSON type makes this possible in sql directly.

But it doesnt end here:
If the extra column on payments was changed to JSON type aswell, the api could support filtering by extra data too. So you could easily get all payments with the tag `boltz`. Something like that isnt possible right now.
If the extra column becomes more viable more extensions could use it to put some metadata alongside the payments, e.g. the discordbot could store the channel id the payment happened in so it can display stats.

**Comment by jackstar12** [comment]
node ui can now be turned on or off in the admin ui:
image

You can optionally expose some info about your node on a public path (/node/public) aswell:
image

**Comment by jackstar12** [comment]
The changes regarding the `services.py` fix type errors which now arise because I've typed the get_wallet_class function correctly

**Comment by arcbtc** [comment]
Getting 
image

**Comment by jackstar12** [comment]
When writing the node tests I also figured out that the wallet class isn't initialized correctly, so `is_fake` was always true. This is now fixed and the `test_pay_real_invoice` test now actually runs (never ran before) and seems to cause some problems.

**Comment by kirito41dd** [comment]
very good job, i very need this feature! when can i try it？

**Comment by jackstar12** [comment]
> very good job, i very need this feature! when can i try it？
if you _really_ want to you can checkout to this branch. Will take a bit more until its fully finished, Ive been extracting some of the changes here into other smaller prs to maoe this one less bulky.

**Comment by callebtc** [comment]
Resolved merge conflicts to new main

**Comment by arcbtc** [comment]
LFGTM!
Its the best node management software on the planet, its not fair the world doesn't have it yet

File: lnbits/app.py [hunk 1]
```diff
@@ -328,6 +328,8 @@ async def lnbits_startup():
             # check extensions after restart
             await check_installed_extensions(app)
 
+            asyncio.create_task(cache.invalidate_forever())
```
Comment (motorina0) [inline lnbits/app.py:331]: - shouldn't this be part on `main` (not part of node management functionality)?

File: lnbits/core/views/generic.py [hunk 2]
```diff
@@ -387,6 +387,44 @@ async def manifest(usr: str):
     }
 
 
+@core_html_routes.get("/node", response_class=HTMLResponse)
+async def node(request: Request, user: User = Depends(check_admin)):
+    if not settings.lnbits_node_ui:
+        raise HTTPException(status_code=HTTPStatus.NOT_FOUND)
```
Comment (motorina0) [inline lnbits/core/views/generic.py:393]: - maybe `HTTP 503 Service Unavailable` is more appropriate
- `404 Not Found` is if the route/resource does not exist

File: lnbits/core/views/node_api.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+)
+@public_node_api.get(
+    "/rank",
+    description="Retrieve node ranks from https://1ml.com",
+    response_model=Optional[NodeRank],
+)
+async def api_get_1ml_stats(node: Node = Depends(require_node)) -> Optional[NodeRank]:
+    node_id = await node.get_id()
+    async with httpx.AsyncClient() as client:
+        # node_id = "026165850492521f4ac8abd9bd8088123446d126f648ca35e60f88177dc149ceb2"
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:179]: - remove dead code

File: lnbits/static/js/components.js [hunk 4]
```diff
@@ -178,6 +178,7 @@ Vue.component('lnbits-extension-list', {
 })
 
 Vue.component('lnbits-admin-ui', {
+  props: ['node'],
```
Comment (motorina0) [inline lnbits/static/js/components.js:181]: - nitpick: this is a `bool`, maybe `has-node` is a more suggestive name

File: lnbits/nodes/base.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+from pydantic import BaseModel
+
+from lnbits.cache import cache
+from lnbits.db import FilterModel, Filters, Page
+
+if TYPE_CHECKING:
+    from lnbits.wallets.base import Wallet
+
+
+class NodeStatsResponse(NamedTuple):
```
Comment (motorina0) [inline lnbits/nodes/base.py:16]: - not used anywhere

File: lnbits/nodes/base.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    amount: int
+    memo: Optional[str]
+    bolt11: str
+    preimage: Optional[str]
+    payment_hash: str
+    paid_at: Optional[int] = None
+    expiry: Optional[int] = None
+
+
+class NodePaymentsResponse(BaseModel):
```
Comment (motorina0) [inline lnbits/nodes/base.py:137]: - not used anywhere

File: lnbits/nodes/base.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    @abstractmethod
+    async def _get_peer_info(self, pubkey: str) -> NodePeerInfo:
+        pass
+
+    async def get_peer_info(self, pubkey: str) -> NodePeerInfo:
+        info = cache.get(f"peers:{pubkey}")
+        if not info:
+            info = await self._get_peer_info(pubkey)
+            if info.last_timestamp:
+                cache.set(f"peers:{pubkey}", info)
```
Comment (motorina0) [inline lnbits/nodes/base.py:195]: - the cache key might be too generic resulting in conflicts.
- something like `node,peers:{pubkey}` would be safer

File: lnbits/nodes/cln.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+    @catch_rpc_errors
+    async def connect_peer(self, uri: str):
+        # https://docs.corelightning.org/reference/lightning-connect
+        try:
+            await self.ln_rpc("connect", uri)
+        except RpcError as e:
+            message = e.error["message"]
+            code = e.error["code"]
+            print("HII")
```
Comment (motorina0) [inline lnbits/nodes/cln.py:73]: - Hello!

File: lnbits/wallets/base.py [hunk 9]
```diff
@@ -1,5 +1,10 @@
+from __future__ import annotations
+
 from abc import ABC, abstractmethod
-from typing import AsyncGenerator, Coroutine, NamedTuple, Optional
+from typing import TYPE_CHECKING, AsyncGenerator, Coroutine, NamedTuple, Optional, Type
+
+if TYPE_CHECKING:
```
Comment (motorina0) [inline lnbits/wallets/base.py:6]: - why this condition, is it to avoid circular dependencies?

File: lnbits/wallets/lndrest.py [hunk 10]
```diff
@@ -74,6 +77,8 @@ async def cleanup(self):
         except RuntimeError as e:
             logger.warning(f"Error closing wallet connection: {e}")
 
+        self.client = httpx.AsyncClient(verify=self.cert, headers=self.auth)
```
Comment (motorina0) [inline lnbits/wallets/lndrest.py:80]: - why add this?

File: lnbits/core/views/node_api.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            status_code=404,
+            detail="Not enabled",
+        )
+    return NODE
+
+
+def check_public():
+    if not (settings.lnbits_node_ui and settings.lnbits_public_node_ui):
+        raise HTTPException(
+            status_code=404,
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:46]: - HTTP 503?

File: lnbits/core/views/node_api.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    status_code=200,
+    dependencies=[Depends(require_node)],
+)
+async def api_get_ok():
+    pass
+
+
+@public_node_api.get("/info", response_model=PublicNodeInfo)
+async def api_get_public_info(node: Node = Depends(require_node)) -> PublicNodeInfo:
+    return await node.get_public_info()
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:69]: - this goes directly to the LN node, can't it be abused (since it is public)?

Comment (motorina0) [reply lnbits/core/views/node_api.py:69]: - same for all `node: Node = Depends(require_node)`

File: lnbits/settings.py [hunk 13]
```diff
@@ -218,6 +220,12 @@ class FundingSourcesSettings(
     lnbits_backend_wallet_class: str = Field(default="VoidWallet")
 
 
+class NodeUISettings(LNbitsSettings):
```
Comment (motorina0) [inline lnbits/settings.py:223]: please add some comments to these new fields

File: lnbits/core/views/node_api.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+@node_api.get("/channels")
+async def api_get_channels(
+    node: Node = Depends(require_node),
+) -> Optional[List[NodeChannel]]:
+    return await node.get_channels()
+
+
+@node_api.post("/channels", response_model=ChannelPoint)
+async def api_create_channel(
+    node: Node = Depends(require_node),
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:88]: - this is a `POST` with no user permission required (anyone can call it if a certain flag is true)?

File: lnbits/core/views/node_api.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    return await node.open_channel(peer_id, funding_amount, push_amount, fee_rate)
+
+
+@node_api.delete("/channels")
+async def api_delete_channel(
+    short_id: Optional[str],
+    funding_txid: Optional[str],
+    output_index: Optional[int],
+    force: bool = False,
+    node: Node = Depends(require_node),
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:103]: - this is a `DELETE` with no user permission required (anyone can call it if a certain flag is true)?

File: lnbits/core/views/node_api.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        ChannelPoint(funding_txid=funding_txid, output_index=output_index)
+        if funding_txid is not None and output_index is not None
+        else None,
+        force,
+    )
+
+
+@node_api.get("/payments", response_model=Page[NodePayment])
+async def api_get_payments(
+    node: Node = Depends(require_node),
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:116]: - is it ok for this info to be accessed by anyone?

File: lnbits/core/views/node_api.py (hunk 14 above)
Comment (jackstar12) [reply lnbits/core/views/node_api.py:88]: The entire `node_api` router is protected by the `check_admin` dependency. It is declared above.

File: lnbits/core/views/node_api.py [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+        raise HTTPException(
+            HTTP_503_SERVICE_UNAVAILABLE,
+            detail="You can enable node transactions in the Admin UI",
+        )
+    return await node.get_payments(filters)
+
+
+@node_api.get("/invoices", response_model=Page[NodeInvoice])
+async def api_get_invoices(
+    node: Node = Depends(require_node),
```
Comment (motorina0) [inline lnbits/core/views/node_api.py:129]: - is it ok for this info to be accessed by anyone?

File: lnbits/core/views/node_api.py (hunk 14 above)
Comment (motorina0) [reply lnbits/core/views/node_api.py:88]: ah, ok!

File: lnbits/core/views/node_api.py (hunk 12 above)
Comment (jackstar12) [reply lnbits/core/views/node_api.py:69]: Its probably a good idea to cache it

File: lnbits/wallets/base.py (hunk 9 above)
Comment (jackstar12) [reply lnbits/wallets/base.py:6]: yes

File: lnbits/core/views/node_api.py (hunk 14 above)
Comment (motorina0) [reply lnbits/core/views/node_api.py:90]: - even so, for some operations (like `create_channel`) it should be reserved for `super_admin`
- the funding source selection & config are reserved for `super_admin`

Comment (jackstar12) [reply lnbits/core/views/node_api.py:90]: if normal admins cant create / close channels, why bother showing them? I'd say either all or nothing

Comment (motorina0) [reply lnbits/core/views/node_api.py:90]: > I'd say either all or nothing
- why?

- normal admins can be non-technical users with special privileges to manage resources
- normal admins already are forbidden from configuring the funding source

**Review by motorina0 — APPROVED** [review]
LGTM, see comments.

Especially: 
- only `super_admin` manages node operations
- others must watch-only (including regular admins)

**Comment by motorina0** [comment]
@dni @jackstar12 do you know why this PR is closed?

**Pull request closed without merging**